src/components_redux/Radios.js

```diff
@@ -3,32 +3,35 @@ import React, { Component } from "react";
 export default class Radios extends Component {
   constructor(props) {
     super(props);
-    this.state = {
-    };
+    this.state = {};
   }
 
   radioChange = event => {
-    this.setState({
-      type: event.target.value
-    });
+    this.setState(
+      {
+        type: event.target.value,
+        selectedValue: event.target.value
+      },
+      () => this.props.handleChangeShow(this.state.selectedValue)
```

Comment:
Can you explain me what `() => this.props.handleChangeShow(this.state.selectedValue)` is doing? And why you do it there?

Reply:
That is callback function. Here we take the selectedValue and pass it to the handleChangeShow in input and depending on the value we set showComponent.

Reply:
That is correct and this is working and in normal javascript would be correct, but in React in this particular case results in multiple render calls.

What happens:
1. You click the radio button and this handler is called
2. You update your state, this tells React to do a re-render to see if some elements need updates
3. After this is finished (as you put it in the callback function of setState) you call handleChangeShow callback, which does again a setState, which results in a second re-render as Radios is part of render of the Input form.

React has some logic internally to merge multiple setStates together and only do one render. But as you explicit say to wait till it finished, these optimizations are not used.

To use this mechanism `this.props.handleChangeShow(this.state.selectedValue)` needs to be called just after setState, without waiting to be finished. BUT then `this.state.selectedValue` can't be used, as the state could be not up-to-date yet. So you would need to use `this.props.handleChangeShow(event.target.value)`.

In this scenario the following will happen:
1. You click the radio button and this handler is called
2. You update your state - setSate is async, but you do not wait now, so it continues with `this.props.handleChangeShow(event.target.value)`, which will also call setState
3. React "merges" both setState calls now and only renders the Input form once.

There are cases where its correct to use the callback function inside setState, like doing a API call. You will also get the complete merged state of all setStates then, which is what you want if you do API calls.
But in this case where you just want to "delegate" a state, it should not be used.

Reply:
I understand everything and fix this in the new commit.